xla/service/collective_quantizer.cc

```diff
@@ -92,9 +95,32 @@ HloInstruction* ApplyUnary(HloInstruction* instr,
   return instr;
 }
 
-// Recursively collects unary, divide, or multiply operands of instr until a
-// conversion to a wider type is reached. Returns an empty vector when no
-// conversion is reached.
+// Returns whether instr is replicated across all partitions associated with
+// module. Returns false when there are multiple replicas.
+absl::StatusOr<bool> InstrIsReplicated(HloModule* module,
+                                       HloInstruction* instr) {
+  // The replication analysis only verifies the replication of instr across
+  // partitions, not replicas. The replica count must be one to ensure instr is
+  // replicated across all devices.
+  if (module->config().replica_count() > 1) {
+    return false;
+  }
+
+  TF_ASSIGN_OR_RETURN(
+      auto replication_analysis,
+      HloReplicationAnalysis::Run(module,
+                                  /*cross_partition_spmd=*/true));
+  if (!replication_analysis->HloInstructionIsReplicatedAt(instr, {})) {
+    return false;
+  }
+
+  return true;
+}
+
+// Recursively collects and returns unary, divide, or multiply operands of instr
+// until a conversion to a wider type is reached. Returns the collected ops in
+// top-down order (operand-to-user). Returns an empty vector when no conversion
```

Comment:
Doesn't this return user-to-operand? It is reversed outside the function call, but this function itself returns user-to-operand.